Source/HTTP/Curl/CurlEasyRequest.cpp

```diff
@@ -0,0 +1,345 @@
+#include "pch.h"
+#include "CurlEasyRequest.h"
+#include "CurlProvider.h"
+
+namespace xbox
+{
+namespace http_client
+{
+
+CurlEasyRequest::CurlEasyRequest(CURL* curlEasyHandle, HCCallHandle hcCall, XAsyncBlock* async)
+    : m_curlEasyHandle{ curlEasyHandle },
+    m_hcCallHandle{ hcCall },
+    m_asyncBlock{ async }
+{
+}
+
+CurlEasyRequest::~CurlEasyRequest()
+{
+    curl_easy_cleanup(m_curlEasyHandle);
+    curl_slist_free_all(m_headers);
+}
+
+Result<HC_UNIQUE_PTR<CurlEasyRequest>> CurlEasyRequest::Initialize(HCCallHandle hcCall, XAsyncBlock* async)
+{
+    CURL* curlEasyHandle{ curl_easy_init() };
+    if (!curlEasyHandle)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "CurlEasyRequest::Initialize:: curl_easy_init failed");
+        return E_FAIL;
+    }
+
+    http_stl_allocator<CurlEasyRequest> a{};
+    HC_UNIQUE_PTR<CurlEasyRequest> easyRequest{ new (a.allocate(1)) CurlEasyRequest{ curlEasyHandle, hcCall, async } };
+
+    // body (first so we can override things curl "helpfully" sets for us)
+    uint8_t const* body = nullptr;
+    uint32_t bodySize = 0;
+    RETURN_IF_FAILED(HCHttpCallRequestGetRequestBodyBytes(hcCall, &body, &bodySize));
+
+    if (bodySize > 0)
+    {
+        // we set both POSTFIELDSIZE and INFILESIZE because curl uses one or the
+        // other depending on method
+        RETURN_IF_FAILED(easyRequest->SetOpt<long>(CURLOPT_POSTFIELDSIZE, bodySize));
+        RETURN_IF_FAILED(easyRequest->SetOpt<long>(CURLOPT_INFILESIZE, bodySize));
+
+        // read callback
+        RETURN_IF_FAILED(easyRequest->SetOpt<curl_read_callback>(CURLOPT_READFUNCTION, &ReadCallback));
+        RETURN_IF_FAILED(easyRequest->SetOpt<void*>(CURLOPT_READDATA, easyRequest.get()));
```

Comment:
Some options that maybe you want to set that I thought LibHttpClient supported with the direct WinHTTP implementation?
CURLOPT_USERAGENT
CURLOPT_ACCEPT_ENCODING

Reply:
I do see that winHttp provider sets libHttpClient as the User-Agent if one wasn't provided by the caller.  If clients manually set User-Agent header and we set CURLOP_USERAGENT, how is that handled?

I don't see where accept encoding is set in the WinHttp implementation, but maybe I'm just missing it.

Reply:
I was pretty sure that WinHTTP will override any specified user agent header regardless of whether you're using xcurl or not.  If it doesn't, then using the existing header mechanism is fine.  I thought that the encoding was set to JSON somewhere for winhttp?

Reply:
I think XSAPI sets the accept encoding header, but its not done anywhere in libHttpClient.  I will add a libHttpClient User-Agent if one was not manually specified (we want to allow XSAPI to set its own User-Agent for example)

Reply:
How would libhc even know what encodings the client accepts?